File: 12_Infix-Postfix/Ideal Solutions/infixPostfix.py
OPERATORS = ['+', '-', '*', '/', '(', ')', '^']
PRIORITY = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}

# Convert Infix Expression to Postfix Expression
def infixPostfix(expression):
    stack = []
    result = ''

    for ch in expression:
        if ch not in OPERATORS:
            result += ch
        elif ch == '(':
            stack.append('(')
        elif ch == ')':
            while stack and stack[-1] != '(':
                result += stack.pop()
        else:
            while stack and stack[-1] != '(' and PRIORITY[ch] <= PRIORITY[stack[-1]]:
                result += stack.pop()
            stack.append(ch)

    while stack:
        result += stack.pop()

    return result


if __name__ == "__main__":

    # Input
    array = [1, 2, 1, 3, 5, 6, 4]
    N = 10
    M = 6

    expression = "a+b-c+d*(e-f)/g+(h*(i/j))"
    print(infixPostfix(expression))

File: 12_Infix-Postfix/Ideal Solutions/infixPostfix.js
/*  
Infix Expression- 
Operators are written in-between their operands. This is the usual way we write expressions.
example : a + b * c + d

Postfix Expression- If the operator appears in the expression after the operands.
example : abc * + d+ 

*/

const isalnum = (value) => {
  const regex = /^[a-zA-Z0-9]+$/;
  return value.match(regex) ? true : false;
};

const priority = (char) => {
  if (char == "^") return 3;
  else if (char == "/" || char == "*") return 2;
  else if (char == "+" || char == "-") return 1;
  else return -1;
};

function infixPostfix(s) {
  let st = []; 
  let result = "";

  for (let i = 0; i < s.length; i++) {
    let c = s[i];
    if (isalnum(c)) result += c;
    else if (c == "(") st.push("(");
    else if (c == ")") {
      while (st[st.length - 1] != "(") {
        result += st[st.length - 1];
        st.pop();
      }
      st.pop();
    } else {
      while (st.length != 0 && priority(s[i]) <= priority(st[st.length - 1])) {
        result += st[st.length - 1];
        st.pop();
      }
      st.push(c);
    }
  }

  while (st.length != 0) {
    result += st[st.length - 1];
    st.pop();
  }

  return result;
}
console.log(infixPostfix("a+b-c+d*(e-f)/g+(h*(i/j))".split("")));
